com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs

```diff
@@ -8,101 +8,362 @@
-        public event NotifyClientAllClientsLoadedSceneDelegate OnNotifyClientAllClientsLoadedScene;
+        internal static bool IsTesting;
+
+        /// <summary>
+        /// The delegate callback definition for scene event notifications
+        /// For more details review over <see cref="SceneEvent"/> and <see cref="SceneEventData"/>
+        /// </summary>
+        /// <param name="sceneEvent"></param>
+        public delegate void SceneEventDelegate(SceneEvent sceneEvent);
+
+        /// <summary>
+        /// Event that will notify the local client or server of all scene events that take place
+        /// For more details review over <see cref="SceneEvent"/>, <see cref="SceneEventData"/>, and <see cref="SceneEventData.SceneEventTypes"/>
+        /// </summary>
+        public event SceneEventDelegate OnSceneEvent;
 
         internal readonly HashSet<string> RegisteredSceneNames = new HashSet<string>();
         internal readonly Dictionary<string, uint> SceneNameToIndex = new Dictionary<string, uint>();
         internal readonly Dictionary<uint, string> SceneIndexToString = new Dictionary<uint, string>();
-        internal readonly Dictionary<Guid, SceneSwitchProgress> SceneSwitchProgresses = new Dictionary<Guid, SceneSwitchProgress>();
-        internal readonly Dictionary<uint, NetworkObject> ScenePlacedObjects = new Dictionary<uint, NetworkObject>();
-
-        private static Scene s_LastScene;
-        private static string s_NextSceneName;
-        private static bool s_IsSwitching = false;
-        internal static uint CurrentSceneIndex = 0;
-        internal static Guid CurrentSceneSwitchProgressGuid = new Guid();
+        internal readonly Dictionary<Guid, SceneEventProgress> SceneEventProgressTracking = new Dictionary<Guid, SceneEventProgress>();
+
+        /// <summary>
+        /// Used to track in-scene placed NetworkObjects
+        /// We store them by:
+        /// [GlobalObjectIdHash][Scene.Handle][NetworkObject]
+        /// The Scene.Handle aspect allows us to distinguish duplicated in-scene placed NetworkObjects created by the loading
+        /// of the same additive scene multiple times.
+        /// </summary>
+        internal readonly Dictionary<uint, Dictionary<int, NetworkObject>> ScenePlacedObjects = new Dictionary<uint, Dictionary<int, NetworkObject>>();
+
+        /// <summary>
+        /// This is used for the deserialization of in-scene placed NetworkObjects in order to distinguish duplicated in-scene
+        /// placed NetworkObjects created by the loading of the same additive scene multiple times.
+        /// </summary>
+        internal Scene SceneBeingSynchronized;
+
+        // Used for observed object synchronization
+        private readonly List<NetworkObject> m_ObservedObjects = new List<NetworkObject>();
+
+        // Used to track which scenes are currently loaded
+        private Dictionary<string, Dictionary<int, Scene>> m_ScenesLoaded = new Dictionary<string, Dictionary<int, Scene>>();
```

Comment:
I have 2 scenes with the same name if they are in different folders.  Could we have an issue keying off of the string scene name?

Reply:
Nope! :)
But... you made me realize I need a better comment here!  (_more awesomeness from your feedback_)
The reason why we won't get confused (based on usage) is that we store them as follows:
[SceneName][SceneHandle][Scene]
So, if you had `Assets/MyScenesA/MyScene` and `Assets/MyScenesB/MyScene` when you loaded each scene they would both have different handles and while the scene names are the same they are distinguished by their handles.

Reply:
what if we were to use `SceneHandle` directly without pre-subsetting it with scene string name/path — wouldn't be simpler and more efficient?

Reply:
You know, I will mull over that... I like the idea but need to do a double take on that code to make sure I don't use the scene name for some other form of validation/check.  If not, then I am ripping out the scene name part for sure.
Good idea!

Reply:
@MFatihMAR 
The most recent commit implements your suggestion.  It did help reduce some of the code complexity in a few methods, and is indeed a more efficient approach.  
👍